x/qgb/orchestrator/orchestrator.go

```diff
@@ -4,95 +4,436 @@ import (
-	evmPrivateKey ecdsa.PrivateKey
-	bridgeID      ethcmn.Hash
+func NewOrchestrator(
+	logger tmlog.Logger,
+	querier Querier,
+	broadcaster Broadcaster,
+	retrier Retrier,
+	signer *paytypes.KeyringSigner,
+	evmPrivateKey ecdsa.PrivateKey,
+) *Orchestrator {
+	orchEthAddr, err := stakingtypes.NewEthAddress(crypto.PubkeyToAddress(evmPrivateKey.PublicKey).Hex())
+	if err != nil {
+		panic(err)
+	}
 
-	// celestia related signing
-	orchestratorAddress sdk.AccAddress
-	orchEthAddress      stakingtypes.EthAddress
+	return &Orchestrator{
+		logger:         logger,
+		signer:         signer,
+		evmPrivateKey:  evmPrivateKey,
+		orchEthAddress: *orchEthAddr,
+		querier:        querier,
+		broadcaster:    broadcaster,
+		retrier:        retrier,
+	}
 }
 
-func (oc *orchestrator) processValsetEvents(ctx context.Context, valsetChannel <-chan types.Valset) error {
-	for valset := range valsetChannel {
-		signBytes, err := valset.SignBytes(oc.bridgeID)
+func (orch Orchestrator) Start(ctx context.Context) {
+	// contains the nonces that will be signed by the orchestrator.
+	noncesQueue := make(chan uint64, 100)
+	defer close(noncesQueue)
+
+	// used to send a signal when the nonces processor wants to notify the nonces enqueuing services to stop.
+	signalChan := make(chan struct{})
+
+	withCancel, cancel := context.WithCancel(ctx)
+
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		err := orch.enqueueMissingEvents(withCancel, noncesQueue, signalChan)
 		if err != nil {
-			oc.logger.Error(fmt.Sprintf("valset nonce %d: %s", valset.Nonce, err.Error()))
-			continue
+			orch.logger.Error("error enqueing missing attestations", "err", err)
+			cancel()
 		}
+		orch.logger.Error("stopping enqueing missing attestations")
+	}()
 
-		signature, err := types.NewEthereumSignature(signBytes.Bytes(), &oc.evmPrivateKey)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		err := orch.startNewEventsListener(withCancel, noncesQueue, signalChan)
 		if err != nil {
-			oc.logger.Error(fmt.Sprintf("valset nonce %d: %s", valset.Nonce, err.Error()))
-			continue
+			orch.logger.Error("error listening to new attestations", "err", err)
+			cancel()
 		}
+		orch.logger.Error("stopping listening to new attestations")
+	}()
 
-		// create and send the valset hash
-		msg := types.NewMsgValsetConfirm(
-			valset.Nonce,
-			oc.orchEthAddress,
-			oc.orchestratorAddress,
-			ethcmn.Bytes2Hex(signature),
-		)
-
-		hash, err := oc.broadcaster.BroadcastTx(ctx, msg)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		err := orch.processNonces(withCancel, noncesQueue, signalChan)
 		if err != nil {
-			oc.logger.Error(fmt.Sprintf("valset nonce %d: %s", valset.Nonce, err.Error()))
-			continue
+			orch.logger.Error("error processing attestations", "err", err)
+			cancel()
+		}
+		orch.logger.Error("stopping processing attestations")
+	}()
+
+	// FIXME should we add  another go routine that keep checking if all the attestations
+	// were signed every 10min for example?
```

Comment:
@evan-forbes What do you think about this?

Reply:
perhaps, yeah. If this is something that can wait, then let's wait

Reply:
Let's leave it for later then. I will create an issue an issue.